# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: uriloader/exthandler/nsIExternalHelperAppService.idl
# Commit: baffa8a7b961
# Full Hash: baffa8a7b9618e9a02e4bd808adb71dead99cce5
# Author: Matt Woodrow <mwoodrow@mozilla.com>
# Date: 2019-11-12 09:43:07
# Regressor Bug: 1589270
# File Overlap Count: 2
# Description:
#   Bug 1589270 - Part 3: Convert nsExternalHelperApp to use BrowsingContext instead of nsIInterfaceRequestor. r=bzbarsky
#   
#   This also converts MaybeCloseWindowHelper, and results in the window close operations being always run in the parent (even without DocumentChannel).
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D49528
# ==============================================================================

diff -r 68bc2ae9b89a -r baffa8a7b961 uriloader/exthandler/nsIExternalHelperAppService.idl
--- a/uriloader/exthandler/nsIExternalHelperAppService.idl	Fri Nov 08 03:01:03 2019 +0000
+++ b/uriloader/exthandler/nsIExternalHelperAppService.idl	Fri Nov 08 04:35:05 2019 +0000
@@ -13,6 +13,12 @@
 interface nsIMIMEInfo;
 interface nsIWebProgressListener2;
 interface nsIInterfaceRequestor;
+webidl BrowsingContext;
+
+[ptr] native ExternalAppHandlerPtr(nsExternalAppHandler);
+%{C++
+class nsExternalAppHandler;
+%}
 
 /**
  * The external helper app service is used for finding and launching
@@ -43,6 +49,30 @@
                                in nsIInterfaceRequestor aContentContext,
                                in boolean aForceSave,
                                [optional] in nsIInterfaceRequestor aWindowContext);
+  
+  /**
+   * Binds an external helper application to a stream listener. The caller
+   * should pump data into the returned stream listener. When the OnStopRequest
+   * is issued, the stream listener implementation will launch the helper app
+   * with this data.
+   * Replaces doContent for native code, and uses BrowsingContext.
+   *
+   * @param aMimeContentType The content type of the incoming data
+   * @param aRequest The request corresponding to the incoming data
+   * @param aContentContext The BrowsingContext that the request was initiated
+   *  by. Used for closing the window if we opened one specifically for this download.
+   * @param aForceSave True to always save this content to disk, regardless of
+   *  nsIMIMEInfo and other such influences.
+   * @param aWindowContext Used in parenting helper app dialogs, usually
+   *  points to the parent browser window. This parameter may be null,
+   *  in which case dialogs will be parented to aContentContext.
+   * @return A nsIStreamListener which the caller should pump the data into.
+   */
+  [noscript] ExternalAppHandlerPtr createListener (in ACString aMimeContentType,
+                                                   in nsIRequest aRequest,
+                                                   in BrowsingContext aContentContext,
+                                                   in boolean aForceSave,
+                                                   [optional] in nsIInterfaceRequestor aWindowContext);
 
   /**
    * Returns true if data from a URL with this extension combination
